### ISSUE TITLE
how to make screenshot comparison?

### ISSUE THREAD

**Issue by andreabisello** [datakurre/robotframework-selenium2screenshots#5]
hi,
any suggestion to make screenshot comparison and make test fail if two screenshot are not equal?

thanks.

**Comment by emanlove** [comment]
This is actually a fairly difficult task to implement if you want a repeatable and reliable test. I recommend that you take a look at [applitools](https://applitools.com/) as a solution. I will note that I have never used their tool nor ever worked on a project that did. I have studied signal/image processing and after listening to Adam Carmi's, the co-founder/VP for R&D for appitools, [talk at Selenium Conf 2015](https://www.youtube.com/watch?v=CHUuLdkFfm0) they seem to have a good handle on this problem. This is also a great talk to see how difficult the problem can be.

That said there is an abandoned project, [Huxley](https://github.com/facebookarchive/huxley/), that was created by Instragram/Facebook which tried to do this or some similar problem. If you search for selenium image comparison you will find several projects, both open source and closed, that try, like Huxley, to do this.

**Comment by datakurre** [comment]
How about combining Sikuli and Selenium with Robot Framework?

**Comment by andreabisello** [comment]
another way is to use needle (https://github.com/bfirsh/needle) on pure python code, and there is a implementation to use with robot ( https://github.com/laurentbristiel/robotframework-needle ) but the implementation works bad (if test fail, robot hangs) and, even i'm able to import python resources in robot framework, i'm not able to use needle in robot framework because something don't work (i got python exception, but i tried only yesterday evening).

my need is to understand if there is some kind of regression making a visual comparison. there are some tools to obtains this, , and applitools is awesome, but if possibile i want to use only robot framework.

**Comment by yuriqp** [comment]
@abioneperhobby did you have any progress on your search. Like you, I need a solution completely integrated with robot framework, because I already have the functional tests with it and I don't want to double the work, but just include visual tests.

**Comment by andreabisello** [comment]
@yuriqp yes i'm using robotframework-sikulilibrary http://rainmanwy.github.io/robotframework-SikuliLibrary/doc/SikuliLibrary.html

**Comment by yuriqp** [comment]
thanks for the reply, I end up writing my own library for using
perceptualdiff (the same image comparison tool Needle uses) as a
robotframework keyword.

**Comment by bbk4163** [comment]
@abioneperhobby , me too tried with needle but stuck due to hang issue. so looking at other options. how r u utilizing sikulilibrary? any resource to check or any guidelines could you please provide

**Comment by datakurre** [comment]
Probably related https://blog.codecentric.de/en/2017/09/robot-framework-compare-images-screenshots/

**Comment by yuriqp** [comment]
@bbk4163 you can try this :https://github.com/blooparksystems/robotframework-imagecompare

**Comment by bbk4163** [comment]
@yuriqp  thanks for sharing u r custom library will check on that for sure..

**Comment by bbk4163** [comment]
@datakurre,  i tried that one too, but i am facing DLL issue due to capabilities mis-match for which i am taking help [here](https://stackoverflow.com/questions/47678713/imagemagick-throwing-dll-error-im-mod-rl-png-dll-the-specified-module-c?noredirect=1#comment82333522_47678713) .. if you know any details pl help.

**Comment by bbk4163** [comment]
Hi, @yuriqp,  i have installed u r library using pip and i can see in the pip list {robotframework-imagecompare (0.1)}. Now i added Compare Images to my sample script but getting error keyword not found error. then tried to add the library like robotframework-imagecompare OR imagecompare no luck.  pl add if i am missing anything here.

**Comment by andreabisello** [comment]
actually i 

1) Capture Page Screenshot with SeleniumLibrary
2) Split the Page Screenshot in a matrix of smaller images with a python tool to help sikuli recognition
3) Use Sikuli to find images in the screen

obtaining A-B regression testing.

**Comment by datakurre** [comment]
@abioneperhobby Actually, if you are able to install OpenCV and its Python bindings onto your environment, you could use it directly from Python. (OpenCV is the computer vision library used by Sikuli.)

Example: https://nbviewer.jupyter.org/github/datakurre/robotkernel/blob/master/examples/OpenCV.ipynb

**Comment by xr55** [comment]
can someone provide the java code to do image comparison and print the mismatches in a seperate file .
